Extrude From Multiple Extruders On an LCD Controller

gralco opened the pull request.

CptanPanic committed:
Add calculation of difference between auto-correct bed level plane, and measured topo map.

Wackerbarth committed:
Compiler defines USBCON for USB devices

gralco committed:
Added menu items to extrude from multiple extruders.

gralco committed:
Added menu items to extrude from multiple extruders.

thinkyhead commented:
I think we want to start labeling the extruders 1-4 instead of 0-3, at least in the user interface. Also I think we should try to minimize string storage. Instead of making 4 extruder strings, just append the number in code.

AnHardt committed:
Z-height fix V2b

here the othe two occourences of `Z_PROBE_OFFSET_FROM_EXTRUDER` are
replaced with `zprobe_zoffset`.

boelle commented:
Please rebase, the code cant be merged as is

thinkyhead committed:
Merge pull request #2269 from Wackerbarth/SwitchToArduinoIDE1.6.4

Switch to Arduino IDE 1.6.4  for Travis testing

thinkyhead committed:
Cleanup of comments & spacing

thinkyhead committed:
Fix comment on M83, add comment to prepare_move

Wackerbarth committed:
Clean up comments and spacing

AnHardt committed:
Reactivate tests for more extruders in Travis

BOARD_ULTIMAKER was not the default any more
- sed failed to change the board
- RAMPS_EFB has no second extruder
- ...
- Added test for bed

Rebased.

Wackerbarth committed:
Merge pull request #2348 from AnHardt/n-Travis-MoreExtruders

Reactivate tests for more extruders in Travis

AnHardt committed:
Add MagoKimbra to the list of contributors

in Development

boelle committed:
Merge pull request #2356 from AnHardt/MK

Add MagoKimbra to the list of contributors

gralco committed:
Fix rebase and fix conflicts

gralco committed:
Fix rebase and fix conflicts

gralco commented:
This commit has been rebased.

Sniffle committed:
z-probe offset fix

verified via 1.0.2 tag that compiles by multiple people,
credit for the fix goes to paulusjacobus for the initial find and
sniffle for the correction so M851 works.

AnHardt committed:
Merge pull request #2361 from Sniffle/Development

z-probe offset fix

AnHardt committed:
Merge pull request #2351 from AnHardt/n-Z-Offset2b

Z-height fix V2b

thinkyhead committed:
Replace "target" with "destination" in M600

- Also fix a bug in `M600` where `Z_MAX_ENDSTOP_INVERTING` is being
used as the index into the `destination[AxisEnum]` array.

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -749,6 +749,51 @@ static void lcd_move_e() {
+  if (LCD_CLICKED) lcd_goto_menu(lcd_move_menu_axis);
+  active_extruder = original_active_extruder;
+}
+#endif // EXTRUDERS > 3
```
Please replace this code with the following to save some program space:

```
#if EXTRUDERS > 1

  static void lcd_move_e(uint8_t e) {
    uint8_t original_active_extruder = active_extruder;
    active_extruder = e;
    if (encoderPosition != 0) {
      current_position[E_AXIS] += float((int)encoderPosition) * move_menu_scale;
      encoderPosition = 0;
      line_to_current(E_AXIS);
      lcdDrawUpdate = 1;
    }
    if (lcdDrawUpdate) lcd_implementation_drawedit(PSTR("Extruder"), ftostr31(current_position[E_AXIS]));
    if (LCD_CLICKED) lcd_goto_menu(lcd_move_menu_axis);
    active_extruder = original_active_extruder;
  }

  static void lcd_move_e1() { lcd_move_e(1); }
  #if EXTRUDERS > 2
    static void lcd_move_e2() { lcd_move_e(2); }
    #if EXTRUDERS > 3
      static void lcd_move_e3() { lcd_move_e(3); }
    #endif // > 3
  #endif // > 2

#endif // EXTRUDERS > 1
```

gralco replied:
Good idea, and I should make the parameter have a default value of 0.

gralco replied:
How is this:

```
static void lcd_move_e(
#if EXTRUDERS > 1
  uint8_t e = 0
#endif
) {
  #if EXTRUDERS > 1
    unsigned short original_active_extruder = active_extruder;
    active_extruder = e;
  #endif
  if (encoderPosition != 0) {
    current_position[E_AXIS] += float((int)encoderPosition) * move_menu_scale;
    encoderPosition = 0;
    line_to_current(E_AXIS);
    lcdDrawUpdate = 1;
  }
  if (lcdDrawUpdate) lcd_implementation_drawedit(PSTR(MSG_MOVE_E), ftostr31(current_position[E_AXIS]));
  if (LCD_CLICKED) lcd_goto_menu(lcd_move_menu_axis);
  #if EXTRUDERS > 1
    active_extruder = original_active_extruder;
  #endif
}

#if EXTRUDERS > 1
  static void lcd_move_e1() { lcd_move_e(1) }
  #if EXTRUDERS > 2
    static void lcd_move_e2() { lcd_move_e(2) }
    #if EXTRUDERS > 3
      static void lcd_move_e3() { lcd_move_e(3) }
    #endif
  #endif
#endif // EXTRUDERS > 1
```

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -761,10 +806,19 @@ static void lcd_move_menu_axis() {
   MENU_ITEM(submenu, MSG_MOVE_X, lcd_move_x);
   MENU_ITEM(submenu, MSG_MOVE_Y, lcd_move_y);
-  if (move_menu_scale < 10.0) {
+//if (move_menu_scale < 10.0) { //Why exclude Z and E from the 10mm and 100mm menus ???
```
I don't know why this has been this way historically, but please submit this change separately and get some feedback first.

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -761,10 +806,19 @@ static void lcd_move_menu_axis() {
+    #endif //EXTRUDERS > 2
+    #if EXTRUDERS > 3
+    MENU_ITEM(submenu, MSG_MOVE_E3, lcd_move_e3);
+    #endif //EXTRUDERS > 3
```
For consistency, you should arrange these lines like so:

```
    #if EXTRUDERS > 1
      MENU_ITEM(submenu, MSG_MOVE_E1, lcd_move_e1);
      #if EXTRUDERS > 2
        MENU_ITEM(submenu, MSG_MOVE_E2, lcd_move_e2);
        #if EXTRUDERS > 3
          MENU_ITEM(submenu, MSG_MOVE_E3, lcd_move_e3);
        #endif // > 3
      #endif // > 2
    #endif //EXTRUDERS > 1
```

thinkyhead committed:
Enable LCD_PIN_RESET if it exists

AnHardt committed:
Merge pull request #2375 from thinkyhead/lcd_pin_reset

Enable LCD_PIN_RESET if it exists

AnHardt committed:
Merge pull request #2374 from thinkyhead/fix_target_M600

Replace "target" with "destination" in M600

AnHardt committed:
Remove code for testing if probing range is to small

becaue it is over-restrictive and superseeded by the now working tests
if the probe-recktangle can be reached. See iss#2366

AnHardt committed:
Correct the sign of zprobe_zoffset for G29

Sign flipped with #2167.
See also comments
https://github.com/MarlinFirmware/Marlin/issues/2040#issuecomment-118568355
and following.

tonnico committed:
Merge pull request #2391 from AnHardt/z-offset_V4

Correct the sign of zprobe_zoffset for G29

thinkyhead committed:
Merge pull request #2390 from AnHardt/iss2366

Remove code for testing if probing range is to small

thinkyhead committed:
Add extra EOL after M111

thinkyhead committed:
Use disable_all_heaters in M111

thinkyhead committed:
Remove unused MSG_OK_B and MSG_OK_T

thinkyhead committed:
Use SERIAL_EOL in SERIAL_PROTOCOLLN macros

thinkyhead committed:
Prevent a hanging SERIAL_ECHO_START in M111

thinkyhead committed:
Call code_seen only once per parameter

thinkyhead committed:
Fixup comments in code_seen

thinkyhead committed:
Merge pull request #2394 from thinkyhead/fixup_m111_etc

Fixup M111, etc.

thinkyhead committed:
RIGIDBOARD / RIGIDPANEL

thinkyhead committed:
RigidBoard include for pins.h

thinkyhead committed:
RigidBot CONTROLLERFAN_PIN in Configuration_adv.h

thinkyhead committed:
Cleanup of pins_RAMPS_13.h

thinkyhead committed:
Use RIGIDBOT_PANEL for consistency with online docs

thinkyhead committed:
Allow N to take a negative argument

thinkyhead committed:
Move note about LiquidTWI2 to Configuration.h

Wackerbarth committed:
Merge pull request #2406 from thinkyhead/panelolu_note

Move note about LiquidTWI2 to Configuration.h

Wackerbarth committed:
Merge pull request #2401 from thinkyhead/rigidbot_support

Rigidbot support

thinkyhead committed:
Also nullify spaces before the checksum asterisk

thinkyhead committed:
Use st_get_position_mm where possible

thinkyhead committed:
Merge pull request #2409 from thinkyhead/command_null_ending

Also nul spaces before the checksum asterisk

thinkyhead committed:
Remove STEPPER_RESET_FIX option

thinkyhead committed:
Add STEPPER_RESET_PIN support

thinkyhead committed:
Move Y_STOP_PIN for Printrboard with SD

Wackerbarth committed:
Merge pull request #2417 from thinkyhead/finish_rigidboard

STEPPER_RESET_PIN support

AnHardt committed:
Remove the additional pin variable in Servo.cpp

as sugested by @c-born
inhttps://github.com/MarlinFirmware/Marlin/issues/1885#issuecomment-92618240

AnHardt committed:
Add Servo::move() to servo.cpp

move(pin, angel) - Sequence of attach(pin), write(angel),
                   if DEACTIVATE_SERVOS_AFTER_MOVE is defined waits SERVO_DEACTIVATION_DELAY, than detaches.

As we have jitter on the servos during moves of the steppers, and detaching them improves this behaviour,
the usual sequence to handle a servo movement is:
attach(pin)
write(angel)
delay(until the servo finished the move)
detache()

Here a function to handle the complete sequence.

AnHardt committed:
Introduce DEACTIVATE_SERVOS_AFTER_MOVE in Configuration.h

DEACTIVATE_SERVOS_AFTER_MOVE is ought to replace PROBE_SERVO_DEACTIVATION_DELAY.
But it will work on all servo moves, not only the probe servo moves.

AnHardt committed:
Activate the new servo.move() function

by replacing the sequences (attach, write, delay, detach), and their conditions with the new function in Marlin.main.cpp
and removing the old configuration in the ABL part of Configuration.h

AnHardt committed:
Deactivate DEACTIVATE_SERVOS_AFTER_MOVE by default

AnHardt committed:
Merge pull request #2402 from thinkyhead/allow_negative_linenum

Allow N to take a negative argument

AnHardt committed:
Merge pull request #2411 from thinkyhead/st_get_mm

Use st_get_position_mm where possible

Commit:
Always calculate extruder temperature PID d term

AnHardt committed:
Merge pull request #2432 from ayberkozgur/Development

Always calculate extruder temperature PID d term

mkile committed:
Slightly updated translation

Slightly updated translation.

mkile committed:
Slightly updated translation

Slightly updated translation.

mkile committed:
Update language_ru.h

mkile committed:
Update language_ru.h

AnHardt committed:
Merge pull request #2443 from mkile/patch-2

Slightly updated translation for language_ru.h

AnHardt committed:
Merge pull request #2433 from Wackerbarth/USBCON

Use USBCON rather than AT90USB for USB devices

AnHardt committed:
Corrected Bed Height vs. Bed Topology for verbosity > 3

and cleaned out some tabs.

Everything else is a rebased PR#2017 as proposed by @CptanPanic

gralco committed:
Several changes per thinkyhead's comments.

gralco committed:
Several changes per thinkyhead's comments.

thinkyhead committed:
Cleanup and dev notes in Marlin.h

thinkyhead committed:
Require temp sensor for bed thermal protection

thinkyhead committed:
Require extruders for thermal code

thinkyhead committed:
Define extern delta only for delta

AnHardt committed:
Eliminate the further use of the pin-parameter

thinkyhead committed:
Merge pull request #2427 from AnHardt/Servo-Move

Anti-jitter for all servo moves

thinkyhead committed:
Update configs for new servo deactivation

thinkyhead committed:
Update configs formatting

thinkyhead committed:
Remove PID examples from preconfigured configs

thinkyhead committed:
Fix lsDive filename stack overrun

AnHardt committed:
Merge pull request #2448 from thinkyhead/update_config_jul15

Update Configurations

thinkyhead committed:
Merge pull request #2444 from AnHardt/pr2017clean

Corrected Bed Height vs. Bed Topology map

thinkyhead commented on Marlin/language_en.h:
```diff
@@ -120,6 +120,15 @@
+#endif
+#ifndef MSG_MOVE_E3
+#define MSG_MOVE_E3                         "Extruder3"
+#endif
```
I think we should define `MSG_MOVE_E` differently with this update, replacing it with `MSG_MOVE_E0`:

```
#if EXTRUDERS < 2
  #define MSG_MOVE_E0                       "Extruder"
#else 
  #define MSG_MOVE_E0                       "Extruder0"
#endif
```

And/or, per my long-ago suggestion, we might want to start using different numbering…

```
#if EXTRUDERS < 2
  #define MSG_MOVE_E0                       "Extruder"
#else 
  #define MSG_MOVE_E0                       "Extruder1"
#endif
#define MSG_MOVE_E1                         "Extruder2"
#define MSG_MOVE_E2                         "Extruder3"
#define MSG_MOVE_E3                         "Extruder4"
```

…as I don't think this will bother most users, even though their boards might have "E0" silkscreened on them.

Wackerbarth replied:
But, rather than having the `#if EXTRUDERS < 2` in every language, you should define it like this in the default (en) file.

```
#ifndef MSG_MOVE_ONLY_E
  #define MSG_MOVE_ONLY_E    "Extruder"
#endif 
#ifndef MSG_MOVE_E
  #define MSG_MOVE_E    "Extruder0"
#endif 
#if EXTRUDERS <2
  #define MSG_MOVE_E0  MSG_MOVE_ONLY_E
#else 
  #define MSG_MOVE_E0 MSG_MOVE_E
#endif
```

thinkyhead committed:
Merge pull request #2425 from AnHardt/Servo-Pin

Remove the additional pin variable in Servo.cpp

Wackerbarth commented on Marlin/language_es.h:
```diff
@@ -42,6 +42,9 @@
 #define MSG_MOVE_E                          "Extrusor"
+#define MSG_MOVE_E1                         "Extruder1"
+#define MSG_MOVE_E2                         "Extruder2"
+#define MSG_MOVE_E3                         "Extruder3"
```
```
#define MSG_MOVE_ONLY_E       "Extrusor"
#define MSG_MOVE_E            "Extrusor0"
#define MSG_MOVE_E1           "Extrusor1"
```

etc.

thinkyhead committed:
Move braces in homeaxis to prevent a hanging else

AnHardt committed:
Merge pull request #2449 from thinkyhead/fix_lsdive_storage

Fix lsDive filename stack overrun
Resolving #2434

AnHardt committed:
Merge pull request #2451 from thinkyhead/fix_homeaxis_braces

Move braces in homeaxis to prevent a hanging else

thinkyhead committed:
Minor tweak to M204 code

thinkyhead committed:
Merge pull request #2457 from thinkyhead/M204_legacy_S_param

Minor tweak to M204 code

thinkyhead committed:
Merge pull request #2445 from thinkyhead/bug_patch_plus

Cleanup & compile error patch

thinkyhead committed:
Merge pull request #2420 from thinkyhead/printrboard_ystop_sd

Move Y_STOP_PIN for Printrboard with SD

thinkyhead committed:
Merge pull request #2395 from thinkyhead/case_insensitive_gcode

Call code_seen only once per parameter

thinkyhead committed:
Move the Z probe as far as needed to trigger the switch

Wackerbarth committed:
Merge pull request #2461 from thinkyhead/probe_deeper

Move the Z probe as far as needed to trigger the switch

gralco committed:
Add conditions for extruders that may or may not be used.

gralco committed:
Add conditions for extruders that may or may not be used.

thinkyhead committed:
Mark unchanging arrays as const

The compiler may be able to optimize if it knows an array won’t be
changing.

thinkyhead committed:
Merge pull request #2494 from thinkyhead/servo_consts

Mark unchanging arrays as const

AnHardt committed:
Merge pull request #2497 from thinkyhead/cleanup_M81

Reduce code in gcode_M81

gralco committed:
Merge branch 'Extrude-From-Multiple-Extruders-LCD' of https://github.com/gralco/Marlin into Extrude-From-Multiple-Extruders-LCD
